_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,598 @@
+
+Below is an example of a type of dashboard we had. When we approached 5kB
+remaining in our firmware image, we knew what to do.
+
+![](/img/tracking-code-size/running-out-code-space.png)
+
+Hopefully this post provides the tools you need to have a dashboard like this
+too!
+
+<!-- excerpt start -->
+
+Code size information should be surfaced and tracked throughout the development
+of an embedded software project, especially if there is even a slight chance of
+running out in the future. In this post, we'll cover why this is the case, show
+how to track code size, and show how to calculate code size deltas on pull
+requests to help keep code size to a minimum.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Why Track Code Size?
+
+If you are a firmware engineer, you've likely run into code sizes issues. If
+not, it is likely because the firmware engineers before you fought for an MCU
+with more micro-flash space or because the company went overboard and bought the
+most expensive chip.
+
+A team that has extra space in the firmware image will find a way to fill it,
+whether it is extra debug logs, a full-featured command-line interface, or fancy
+animations or images. With this in mind, it would be wise to constantly track
+and measure the code size of a project during development to know when and why
+the code size grows. This would mean that each commit and a handful of build
+configurations should be tracked so an accurate picture can be painted.
+
+I've seen many projects start to have a print out like below after each build,
+which is nice, but it isn't enough to keep the code size from being gobbled up.
+
+```
+Memory region         Used Size  Region Size  %age Used
+           FLASH:      153108 B         1 MB     14.60%
+            SRAM:       44544 B       256 KB     16.99%
+        IDT_LIST:          56 B         2 KB      2.73%
+```
+
+Once you have the code size for each commit and build combination, you and your
+entire team can:
+
+- Prevent large (possibly accidental) code size increases due to non-optimal
+  code for function usage (e.g. a few years ago, by adding a call to newlib's
+  `sscanf(...)`, I increased the firmware size by 2kB).
+- You can retroactively find large spikes or decreases in code size when looking
+  for easy-win optimizations.
+- Estimate when the code size will exceed the limit by looking at historical
+  data.
+- Over time, learn which coding patterns contribute to the code size increases.
+- Calculate code size deltas on every new pull request. You would likely prefer
+  to merge a shiny new feature which adds 2kB of code instead of a poorly
+  implemented bug-fix which also adds 2kB.
+
+## How to Track Code Size
+
+In the remainder of the blog post, we'll go through an example of tracking the
```

Comment:
In the remainder of this blog post, we will set up a code-size tracking dashboard for the popular open-source RTOS...